src/adapter/src/catalog/open.rs

```diff
@@ -1769,6 +1783,95 @@ impl Catalog {
+
+    for builtin_cluster in BUILTIN_CLUSTERS {
+        if !cluster_names.contains(builtin_cluster.name) {
+            let id = txn.get_and_increment_id(SYSTEM_CLUSTER_ID_ALLOC_KEY.to_string())?;
+            let id = ClusterId::System(id);
+            txn.insert_system_cluster(
+                id,
+                builtin_cluster.name,
+                vec![],
+                builtin_cluster.privileges.to_vec(),
+                mz_catalog::durable::ClusterConfig {
+                    // TODO: Should builtin clusters be managed or unmanaged?
+                    variant: mz_catalog::durable::ClusterVariant::Unmanaged,
+                },
+            )?;
+        }
+    }
+    Ok(())
+}
+
+fn add_new_builtin_cluster_replicas_migration(
+    txn: &mut mz_catalog::durable::Transaction<'_>,
+    builtin_cluster_replica_size: String,
+) -> Result<(), mz_catalog::durable::CatalogError> {
+    let cluster_lookup: BTreeMap<_, _> = txn
+        .get_clusters()
+        .map(|cluster| (cluster.name, cluster.id))
+        .collect();
+
+    let replicas: BTreeMap<_, _> =
+        txn.get_cluster_replicas()
+            .fold(BTreeMap::new(), |mut acc, replica| {
+                acc.entry(replica.cluster_id)
+                    .or_insert_with(BTreeSet::new)
+                    .insert(replica.name);
+                acc
+            });
+
+    for builtin_replica in BUILTIN_CLUSTER_REPLICAS {
+        let cluster_id = cluster_lookup
+            .get(builtin_replica.cluster_name)
+            .expect("builtin cluster replica references non-existent cluster");
+
+        let replica_names = replicas.get(cluster_id);
+        if matches!(replica_names, None)
+            || matches!(replica_names, Some(names) if !names.contains(builtin_replica.name))
+        {
+            let replica_id = txn.get_and_increment_id(SYSTEM_REPLICA_ID_ALLOC_KEY.to_string())?;
+            let replica_id = ReplicaId::System(replica_id);
+            let config = builtin_cluster_replica_config(builtin_cluster_replica_size.clone());
+            txn.insert_cluster_replica(
+                *cluster_id,
+                replica_id,
+                builtin_replica.name,
+                config,
+                MZ_SYSTEM_ROLE_ID,
+            )?;
+        }
+    }
+    Ok(())
+}
+
+pub(crate) fn builtin_cluster_replica_config(
+    builtin_cluster_replica_size: String,
+) -> mz_catalog::durable::ReplicaConfig {
+    mz_catalog::durable::ReplicaConfig {
+        location: mz_catalog::durable::ReplicaLocation::Managed {
+            availability_zone: None,
+            billed_as: None,
+            disk: false,
+            internal: false,
+            size: builtin_cluster_replica_size,
+        },
+        logging: default_logging_config(),
+        idle_arrangement_merge_effort: None,
+    }
+}
+
+fn default_logging_config() -> ReplicaLogging {
+    ReplicaLogging {
+        log_logging: false,
+        interval: Some(Duration::from_secs(1)),
+    }
+}
```

Comment:
This is mostly code movement. The one difference is that `builtin_cluster_replica_config` and `add_new_builtin_cluster_replicas_migration` take a `String` argument instead of a `BootstrapArgs` argument.